### PR TITLE
Actuate IMC with Torque mode

### PR DESCRIPTION
**Pull request by Ishadijcks**
## Introduction
This PR replaces position mode on the IMC with torque mode. As we still might need position mode in the future, we implemented a system that lets you easily switch between those two modes.
At the bottom is a list of todo's that need to be implemented before this can be merged to develop.

## Features
- Configure the IMC to use Cyclic Torque mode instead of Cyclic Position Mode.
- Send current directly to the IMC instead of just a target position.
- Limits are working for both Torque Mode and Position Mode.

## Todo's before merge
- [x] Implement a mode flag for the IMC to toggle between position and torque (with enums)
- [x] Add this to the hardwarebuilder (see #33 for example)
- [x] Add tests to the hardwarebuilder (#33)
- [x] Write mode to IMC based on object mode
- [x] Add a guard when calling `actuateRad` in torque mode and vice versa.
- [x] Implement tests for this behaviour.
- [x] Change the hardware_interface to `actuateRad()` on position mode and `actuateCurrent()` on Torque mode
- [x] Test on the M3
- [x] Detect it when using an effortcontroller on a joint that is in Position mode and vice versa
- [x] Refactor `actuateCurrent()` from float to int
- [x] Add more MISO variables (motorcurrent, what else?)
- [x] Remove unnecessary `<rosparam param="pid/publish_state">true</rosparam>`

### CODE REVIEW

File: march_hardware/include/march_hardware/ActuationMode.h [hunk 1]
```diff
@@ -60,14 +61,18 @@ class ActuationMode
...
+  std::string toString() const
+  {
+    switch (this->value)
+    {
+      case torque:
+        return "torque";
+      case position:
+        return "position";
+      default:
+        return "unknown";
```
Comment (Ishadijcks) [inline march_hardware/include/march_hardware/ActuationMode.h:73]: Should this throw a warning?

File: march_hardware/include/march_hardware/Joint.h [hunk 2]
```diff
@@ -14,7 +14,7 @@ class Joint
 {
 private:
   std::string name;
-  //TODO(bart) review this comment if it still makes sense
+  // TODO(bart) review this comment if it still makes sense
```
Comment (Ishadijcks) [inline march_hardware/include/march_hardware/Joint.h:21]: Does it?

**Comment by Ishadijcks** [comment]
Why is the linter not failing this build?

File: march_hardware_builder/test/yaml/joint/joint_correct_1.yaml [hunk 3]
```diff
@@ -1,3 +1,5 @@
+actuationMode: torque
+
```
Comment (Ishadijcks) [inline march_hardware_builder/test/yaml/joint/joint_correct_1.yaml:2]:
```suggestion
```

File: march_hardware/src/IMotionCube.cpp [hunk 4]
```diff
@@ -61,14 +65,14 @@ void IMotionCube::validateMosiPDOs()
...
-void IMotionCube::writeInitialSettings(uint8 ecatCycleTime)
+void IMotionCube::writeInitialSettings(uint8 ecatCycleTime, uint8_t modeofOp)
 {
   ROS_DEBUG("IMotionCube::writeInitialSettings");
   bool success = true;
   // sdo_bit32(slaveIndex, address, subindex, value);
   // mode of operation
-  success &= sdo_bit8(slaveIndex, 0x6060, 0, 8);
-
+  ROS_INFO("the mode of operation is: %i", modeofOp);
```
Comment (martijnvandermarel) [inline march_hardware/src/IMotionCube.cpp:74]: If you want to keep this as an info print, please add the joint name (e.g. "Mode of operation for left_knee is torque"). Since the IMC doesn't have the joint name, you probably need to move the print to the joint class

File: march_hardware/include/march_hardware/ActuationMode.h [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    {
+      this->value = position;
+    }
+    else if (actuationMode == "torque")
+    {
+      this->value = torque;
+    }
+    else
+    {
+      //      ROS_ASSERT_MSG(false, "Unknown actuation mode %s, setting to unknown mode", actuationMode.c_str());
```
Comment (Ishadijcks) [inline march_hardware/include/march_hardware/ActuationMode.h:32]: Commented code

File: march_hardware/include/march_hardware/ActuationMode.h (hunk 1 above)
Comment (Ishadijcks) [reply march_hardware/include/march_hardware/ActuationMode.h:73]: Still relevant 2 months later

File: march_hardware/include/march_hardware/Joint.h (hunk 2 above)
Comment (Ishadijcks) [reply march_hardware/include/march_hardware/Joint.h:21]: Still relevant 2 months later

File: march_hardware/include/march_hardware/Joint.h [hunk 6]
```diff
@@ -17,28 +18,41 @@ class Joint
...
-  Joint(std::string name, bool allowActuation, IMotionCube iMotionCube);
-  Joint(std::string name, bool allowActuation, IMotionCube iMotionCube, int netNumber);
+  // TODO(Tim) pass by reference or pointer instead of making copy
+  Joint(std::string name, bool allowActuation, TemperatureGES temperatureGES, IMotionCube iMotionCube,
+        std::string actuationmode);
+  Joint(std::string name, bool allowActuation, TemperatureGES temperatureGES, IMotionCube iMotionCube, int netNumber,
+        std::string actuationmode);
+  Joint(std::string name, bool allowActuation, TemperatureGES temperatureGES, std::string actuationmode);
+  Joint(std::string name, bool allowActuation, IMotionCube iMotionCube, std::string actuationmode);
+  Joint(std::string name, bool allowActuation, IMotionCube iMotionCube, int netNumber, std::string actuationmode);
```
Comment (Ishadijcks) [inline march_hardware/include/march_hardware/Joint.h:42]: This should be refactored to using proper initialization lists.

File: march_hardware/include/march_hardware/Joint.h [hunk 7]
```diff
@@ -17,28 +18,41 @@ class Joint
...
 
   void initialize(int ecatCycleTime);
   void prepareActuation();
   // TODO(Martijn) Refactor this to make joint less dependent on knowledge of the IMC
   void resetIMotionCube();
 
   void actuateRad(float targetPositionRad);
+  void actuateCurrent(int targetCurrentRad);
```
Comment (Ishadijcks) [inline march_hardware/include/march_hardware/Joint.h:50]: CurrentRad?

File: march_hardware/src/EtherCAT/EthercatMaster.cpp [hunk 8]
```diff
@@ -97,7 +97,8 @@ void EthercatMaster::start()
   {
     // All slaves in operational state
     ROS_INFO("Operational state reached for all slaves");
-    isOperational = true;
+
+      isOperational = true;
```
Comment (Ishadijcks) [inline march_hardware/src/EtherCAT/EthercatMaster.cpp:101]: Unneeded change

File: march_hardware/src/EtherCAT/EthercatSDO.cpp [hunk 9]
```diff
@@ -12,19 +12,19 @@ namespace march4cpp
 // TODO(Isha, Martijn, Tim) refactor this with more generic types
 int sdo_bit8(int slave, uint32_t index, uint8_t sub, uint8_t value)
 {
-  ROS_DEBUG("sdo_bit8: slaveIndex %i, reg 0x%X, subindex %i, value 0x%X", slave, index, sub, value);
+  ROS_INFO("sdo_bit8: slaveIndex %i, reg 0x%X, subindex %i, value 0x%X", slave, index, sub, value);
```
Comment (Ishadijcks) [inline march_hardware/src/EtherCAT/EthercatSDO.cpp:15]: This is a debug

File: march_hardware/src/EtherCAT/EthercatSDO.cpp [hunk 10]
```diff
@@ -12,19 +12,19 @@ namespace march4cpp
...
   return ec_SDOwrite(slave, index, sub, FALSE, 1, &value, EC_TIMEOUTRXM);
 }
 
 int sdo_bit16(int slave, uint32_t index, uint8_t sub, uint16_t value)
 {
-  ROS_DEBUG("sdo_bit16: slaveIndex %i, reg 0x%X, subindex %i, value 0x%X", slave, index, sub, value);
+    ROS_INFO("sdo_bit16: slaveIndex %i, reg 0x%X, subindex %i, value 0x%X", slave, index, sub, value);
```
Comment (Ishadijcks) [inline march_hardware/src/EtherCAT/EthercatSDO.cpp:21]: This is a debug

Formatting

File: march_hardware/src/EtherCAT/EthercatSDO.cpp [hunk 11]
```diff
@@ -12,19 +12,19 @@ namespace march4cpp
...
   return ec_SDOwrite(slave, index, sub, FALSE, 2, &value, EC_TIMEOUTRXM);
 }
 
 int sdo_bit32(int slave, uint32_t index, uint8_t sub, uint32_t value)
 {
-  ROS_DEBUG("sdo_bit32: slaveIndex %i, reg 0x%X, subindex %i, value 0x%X", slave, index, sub, value);
+    ROS_INFO("sdo_bit32: slaveIndex %i, reg 0x%X, subindex %i, value 0x%X", slave, index, sub, value);
```
Comment (Ishadijcks) [inline march_hardware/src/EtherCAT/EthercatSDO.cpp:27]: This is a debug

Formatting

File: march_hardware/src/IMotionCube.cpp [hunk 12]
```diff
@@ -10,19 +10,20 @@
...
 
-void IMotionCube::writeInitialSDOs(int ecatCycleTime)
+void IMotionCube::writeInitialSDOs(int ecatCycleTime, ActuationMode mode)
 {
   mapMisoPDOs();
   mapMosiPDOs();
   validateMisoPDOs();
   validateMosiPDOs();
-  writeInitialSettings(ecatCycleTime);
+  this->actuationMode = mode;
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp:25]: Use initialization list here

File: march_hardware/src/IMotionCube.cpp [hunk 13]
```diff
@@ -32,6 +33,8 @@ void IMotionCube::mapMisoPDOs()
   PDOmap pdoMapMISO = PDOmap();
   pdoMapMISO.addObject(IMCObjectName::StatusWord);      // Compulsory!
   pdoMapMISO.addObject(IMCObjectName::ActualPosition);  // Compulsory!
+  pdoMapMISO.addObject(IMCObjectName::ActualTorque);
+  pdoMapMISO.addObject(IMCObjectName::TargetCurrent);
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp:37]: Why is 1 called Torque and the other current?

File: march_hardware/src/IMotionCube.cpp [hunk 14]
```diff
@@ -44,6 +47,7 @@ void IMotionCube::mapMosiPDOs()
   PDOmap pdoMapMOSI = PDOmap();
   pdoMapMOSI.addObject(IMCObjectName::ControlWord);  // Compulsory!
   pdoMapMOSI.addObject(IMCObjectName::TargetPosition);
+  //  pdoMapMOSI.addObject(IMCObjectName::TargetCurrent);
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp:50]: Commented code

File: march_hardware/src/IMotionCube.cpp [hunk 15]
```diff
@@ -83,7 +87,7 @@ void IMotionCube::writeInitialSettings(uint8 ecatCycleTime)
   success &= sdo_bit32(slaveIndex, 0x607D, 2, this->encoder.getMaxPositionIU());
 
   // Quick stop option
-  success &= sdo_bit16(slaveIndex, 0x605A, 0, 6);
+  //  success &= sdo_bit16(slaveIndex, 0x605A, 0, 6);
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp:90]: Commented code

File: march_hardware/src/IMotionCube.cpp [hunk 16]
```diff
@@ -106,6 +113,31 @@ void IMotionCube::actuateRad(float targetRad)
   this->actuateIU(this->encoder.RadtoIU(targetRad));
 }
 
+void IMotionCube::actuateCurrent(float targetCurrent)
+{
+  ROS_ASSERT_MSG(this->actuationMode == ActuationMode::torque, "trying to actuate current, while actuationmode = "
+                                                               "%s",
+                 this->actuationMode.toString().c_str());
+
+  ROS_ASSERT_MSG(targetCurrent < 20000, "Current of %f is too high.", targetCurrent);
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp]: Magic number

File: march_hardware/src/IMotionCube.cpp [hunk 17]
```diff
@@ -106,6 +113,31 @@ void IMotionCube::actuateRad(float targetRad)
...
+
+  union bit16 targetCurrentStruct;
+  targetCurrentStruct.i = targetCurrent;
+
+  if (this->mosiByteOffsets.count(IMCObjectName::TargetCurrent) != 1)
+  {
+    ROS_WARN("TargetCurrent not defined in PDO mapping, so can't do actuateIU");
+    return;
+  }
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp:131]: Sanity checks should come before any logic (e.g. creating objects)

File: march_hardware/src/IMotionCube.cpp [hunk 18]
```diff
@@ -106,6 +113,31 @@ void IMotionCube::actuateRad(float targetRad)
...
+
+  int targetCurrentLocation = this->mosiByteOffsets[IMCObjectName::TargetCurrent];
+
+  ROS_DEBUG("Trying to actuate slave %d, soem location %d with targetcurrent%i", this->slaveIndex,
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp:135]:
```suggestion
  ROS_DEBUG("Trying to actuate slave %d, soem location %d with targetcurrent %i", this->slaveIndex,
```

File: march_hardware/src/IMotionCube.cpp [hunk 19]
```diff
@@ -106,6 +113,31 @@ void IMotionCube::actuateRad(float targetRad)
   this->actuateIU(this->encoder.RadtoIU(targetRad));
 }
 
+void IMotionCube::actuateCurrent(float targetCurrent)
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp]: Is this a float?

File: march_hardware/src/IMotionCube.cpp [hunk 20]
```diff
@@ -106,6 +113,31 @@ void IMotionCube::actuateRad(float targetRad)
...
+{
+  ROS_ASSERT_MSG(this->actuationMode == ActuationMode::torque, "trying to actuate current, while actuationmode = "
+                                                               "%s",
+                 this->actuationMode.toString().c_str());
+
+  ROS_ASSERT_MSG(targetCurrent < 20000, "Current of %f is too high.", targetCurrent);
+
+  union bit16 targetCurrentStruct;
+  targetCurrentStruct.i = targetCurrent;
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp:125]: Storing a float as int

File: march_hardware/src/Joint.cpp [hunk 21]
```diff
@@ -7,33 +7,48 @@
...
+  , netNumber(netNumber)
+  , actuationMode(ActuationMode(actuationModeName))
 {
   this->name = std::move(name);
   this->allowActuation = allowActuation;
 }
 
-Joint::Joint(std::string name, bool allowActuation, TemperatureGES temperatureGES) : temperatureGES(temperatureGES)
+// Only Temperature GES
+// TODO(bart): make sure these constructors make sense with actuationmode
```
Comment (Ishadijcks) [inline march_hardware/src/Joint.cpp:32]: Does it?

File: march_hardware/src/Joint.cpp [hunk 22]
```diff
@@ -7,33 +7,48 @@
 
 namespace march4cpp
 {
-Joint::Joint(std::string name, bool allowActuation, TemperatureGES temperatureGES, IMotionCube iMotionCube)
-  : temperatureGES(temperatureGES), iMotionCube(iMotionCube)
+// ImotionCube and Temperature GES
+Joint::Joint(std::string name, bool allowActuation, TemperatureGES temperatureGES, IMotionCube iMotionCube,
+             std::string actuationModeName)
+  : temperatureGES(temperatureGES), iMotionCube(iMotionCube), actuationMode(ActuationMode(actuationModeName))
```
Comment (Ishadijcks) [inline march_hardware/src/Joint.cpp:13]: Also be refactored with setters

File: march_hardware/src/Joint.cpp [hunk 23]
```diff
@@ -83,6 +98,17 @@ void Joint::actuateRad(float targetPositionRad)
   this->iMotionCube.actuateRad(targetPositionRad);
 }
 
+void Joint::actuateCurrent(int targetCurrent)
+{
+  // TODO(BaCo) check that the position is allowed and does not exceed (torque) limits.
```
Comment (Ishadijcks) [inline march_hardware/src/Joint.cpp:103]: Does it?

File: march_hardware/src/Joint.cpp [hunk 24]
```diff
@@ -83,6 +98,17 @@ void Joint::actuateRad(float targetPositionRad)
   this->iMotionCube.actuateRad(targetPositionRad);
 }
 
+void Joint::actuateCurrent(int targetCurrent)
```
Comment (Ishadijcks) [inline march_hardware/src/Joint.cpp:101]: Here current is an int?

File: march_hardware/src/PDOmap.cpp [hunk 25]
```diff
@@ -35,70 +35,106 @@ std::map<enum IMCObjectName, int> PDOmap::map(int slaveIndex, enum dataDirection
...
+        reg++;
+      }
+      // else, check if register is full
+      else if (sizeleft <= 0)
+      {
+        sdo_bit32(slaveIndex, reg, 0, counter);
+        reg++;
+        counter = 0;
+        sizeleft = this->bitsPerReg;
+      }
```
Comment (Ishadijcks) [inline march_hardware/src/PDOmap.cpp:78]: What are these changes? Just indentation?

File: march_hardware_interface/config/march3/controllers.yaml [hunk 26]
```diff
@@ -10,11 +10,18 @@ march:
...
         - left_knee
         - right_ankle
         - left_ankle
+      gains: # Required because we're controlling an effort interface
+        right_hip: {p: 50000, i: 5000, d: 5000, i_clamp: 10, publish_state: true, antiwindup: true}
+        left_hip: {p: 50000, i: 5000, d: 5000, i_clamp: 10, publish_state: true, antiwindup: true}
+        right_knee: {p: 50000, i: 5000, d: 5000, i_clamp: 10, publish_state: true, antiwindup: true}
+        left_knee: {p: 50000, i: 5000, d: 5000, i_clamp: 10, publish_state: true, antiwindup: true}
+        right_ankle: {p: 5000, i: 500, d: 5000, i_clamp: 10, publish_state: true, antiwindup: true}
+        left_ankle: {p: 5000, i: 500, d: 5000, i_clamp: 10, publish_state: true, antiwindup: true}
```
Comment (Ishadijcks) [inline march_hardware_interface/config/march3/controllers.yaml:27]: ?

File: march_hardware_interface/config/march4/controllers.yaml [hunk 27]
```diff
@@ -9,12 +9,23 @@ march:
     temperature_sensor:
       type: march_temperature_sensor_controller/MarchTemperatureSensorController
       publish_rate: 1
-    trajectory:
-      type: position_controllers/JointTrajectoryController
+    trajectory_effort:
+      type: effort_controllers/JointTrajectoryController
```
Comment (Ishadijcks) [inline march_hardware_interface/config/march4/controllers.yaml:13]: Just add another controller instead of removing the old one

File: march_hardware_interface/launch/march4.launch [hunk 28]
```diff
@@ -9,7 +9,8 @@
           args="
                 /march/controller/joint_state
                 /march/controller/temperature_sensor
-                /march/controller/trajectory
+                /march/controller/trajectory_effort
+                /march/controller/trajectory_position
```
Comment (Ishadijcks) [inline march_hardware_interface/launch/march4.launch]: Why launch both?

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 29]
```diff
@@ -138,18 +152,32 @@ void MarchHardwareInterface::init()
...
-    JointHandle jointPositionHandle(jointStateHandle, &joint_position_command_[i]);
-
     // Retrieve joint (soft) limits from the urdf
     JointLimits limits;
     getJointLimits(model.getJoint(joint.getName()), limits);
 
-    // Create joint limit interface
-    PositionJointSoftLimitsHandle jointLimitsHandle(jointPositionHandle, limits, soft_limits_[i]);
-    positionJointSoftLimitsInterface.registerHandle(jointLimitsHandle);
+    // TODO(Jitske): make sure that both of the controllers are loaded, but only one controller starts. Currently both
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp:163]: Does it?

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 30]
```diff
@@ -247,17 +272,30 @@ void MarchHardwareInterface::read(ros::Duration elapsed_time)
...
+      if (singleJoint.getActuationMode() == ActuationMode::position)
+      {
+        positionJointSoftLimitsInterface.enforceLimits(elapsed_time);
+        singleJoint.actuateRad(static_cast<float>(joint_position_command_[i]));
+      }
+      else if (singleJoint.getActuationMode() == ActuationMode::torque)
+      {
+        effortJointSoftLimitsInterface.enforceLimits(elapsed_time);
+        singleJoint.actuateCurrent(static_cast<int>(joint_effort_command_[i]));
+      }
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp:308]: Shouldn't all limits be enforced anyways?

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 31]
```diff
@@ -247,17 +272,30 @@ void MarchHardwareInterface::read(ros::Duration elapsed_time)
...
+
+      ROS_INFO("After effort limit: Trying to actuate joint %s, to %lf "
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp]: Not an info

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 32]
```diff
@@ -362,7 +402,8 @@ void MarchHardwareInterface::updatePowerNet()
     catch (std::exception& exception)
     {
       ROS_ERROR("%s", exception.what());
-      ROS_WARN("Reset power net command, in attempt to prevent this exception is thrown again");
+      ROS_WARN("Reset power net command, in attempt to prevent this exception "
+               "is thrown again");
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp:425]: All unneeded changes

File: march_hardware_interface/src/march_hardware_interface.cpp [hunk 33]
```diff
@@ -436,5 +477,35 @@ void MarchHardwareInterface::outsideLimitsCheck(int joint_index)
       throw ::std::runtime_error(errorStream.str());
     }
   }
+
+  if (joint_velocity_[joint_index] > joint_limits_[joint_index].max_velocity)
```
Comment (Ishadijcks) [inline march_hardware_interface/src/march_hardware_interface.cpp]: This should really be another check

**Review by Ishadijcks — COMMENTED** [review]
Quite a lot of comments.

- Please clean up your pr before submitting it. Nobody likes reviewing formatting errors and commented code.
- Use initializer lists with setters instead of 10 constructors for each configuration
- Why is the actuation mode not optional? With unknown as the default. This removes the need to edit all previously existing tests
- Add more tests

File: march_hardware/src/IMotionCube.cpp (hunk 14 above)
Comment (martijnvandermarel) [reply march_hardware/src/IMotionCube.cpp:50]: How does this even work. TargetCurrent should be compulsory as mosi pdo?!? As miso I don't see the added value.

File: march_hardware/src/IMotionCube.cpp (hunk 13 above)
Comment (martijnvandermarel) [reply march_hardware/src/IMotionCube.cpp:37]: Remove TargetCurrent as miso? It's only necessary as mosi. ActualTorque is what you want to read as miso pdo

Comment (martijnvandermarel) [reply march_hardware/src/IMotionCube.cpp:37]: Re @Ishadijcks : ActualTorque and TargetCurrent are the names as defined by Technosoft. Not much we can do about their inconsistency.

File: march_hardware/src/IMotionCube.cpp (hunk 20 above)
Comment (martijnvandermarel) [reply march_hardware/src/IMotionCube.cpp:125]: Yeah if it really is a float, this won't work. You screw up the variable if you interpret a floating point variable as an int.

File: march_hardware/src/PDOmap.cpp (hunk 25 above)
Comment (martijnvandermarel) [reply march_hardware/src/PDOmap.cpp:78]: Nope it seems it still uses hardcoded pdo mapping. Please go back to the dynamic pdo mapping as defined in develop and check if that works.

Comment (martijnvandermarel) [reply march_hardware/src/PDOmap.cpp:78]: This explains why torque mode even works despite not adding it as a mosi pdo object in the imc class, btw.

File: march_hardware/include/march_hardware/Joint.h (hunk 2 above)
Comment (JitskedeVries) [reply march_hardware/include/march_hardware/Joint.h:21]: Changed it a little bit. Still makes sense I think, if you want to explain why they are both there. However, I think the code and the names speak for itself. So maybe there is no comment needed. Any thoughts @Bart-van-Ingen ?

File: march_hardware/src/IMotionCube.cpp (hunk 14 above)
Comment (JitskedeVries) [reply march_hardware/src/IMotionCube.cpp:50]: It is defined in the PDO mapping, should add it again here and delete it in the PDO mapping.

File: march_hardware/src/IMotionCube.cpp (hunk 4 above)
Comment (JitskedeVries) [reply march_hardware/src/IMotionCube.cpp:74]: moved to the joint.cpp

File: march_hardware/src/IMotionCube.cpp (hunk 15 above)
Comment (JitskedeVries) [reply march_hardware/src/IMotionCube.cpp:90]: We removed the code now, because this object can not be mapped.

File: march_hardware/src/IMotionCube.cpp (hunk 19 above)
Comment (JitskedeVries) [reply march_hardware/src/IMotionCube.cpp]: Changed into an int

File: march_hardware/src/IMotionCube.cpp (hunk 16 above)
Comment (JitskedeVries) [reply march_hardware/src/IMotionCube.cpp]: Ideally we would want to implement the max_effort from the urdf. But I could not figure out a good way to implement that. I'll come and talk about this, maybe you have better ideas.

Comment (JitskedeVries) [reply march_hardware/src/IMotionCube.cpp]: Isn't the 0.27 difference from the actuateposition also a magic number?

File: march_hardware/src/IMotionCube.cpp (hunk 20 above)
Comment (JitskedeVries) [reply march_hardware/src/IMotionCube.cpp:125]: targetCurrent is an int now

File: march_hardware/src/IMotionCube.cpp (hunk 17 above)
Comment (JitskedeVries) [reply march_hardware/src/IMotionCube.cpp:131]: Are you referring to the fact that above this code snippet an object is made: targetCurrentStruct? So we should place the sanity check at the beginning of the method (IMotionCube::actuateCurrent)? Or should a seperate method be made in which the mosibyteoffset is checked before ever sending an actuation command?

File: march_hardware/include/march_hardware/Joint.h (hunk 6 above)
Comment (JitskedeVries) [reply march_hardware/include/march_hardware/Joint.h:42]: two questions:

1. with refactoring are you referring to the partial initializer lists used by the above mentioned methods in the joint.cpp file? here some variables are being initialized through an initialization list at the start of the method while others are being set by the body of the function. Should the latter be formatted so that they too are implemented in the initialization list?

2. Do you think it appropriate to use an initialization list in the header? This will remove the functions from the corresponding cpp file.

File: march_hardware/src/IMotionCube.cpp [hunk 34]
```diff
@@ -113,13 +113,13 @@ void IMotionCube::actuateRad(float targetRad)
...
 
-void IMotionCube::actuateCurrent(float targetCurrent)
+void IMotionCube::actuateCurrent(int targetCurrent)
 {
   ROS_ASSERT_MSG(this->actuationMode == ActuationMode::torque, "trying to actuate current, while actuationmode = "
                                                                "%s",
                  this->actuationMode.toString().c_str());
 
-  ROS_ASSERT_MSG(targetCurrent < 20000, "Current of %f is too high.", targetCurrent);
+  ROS_ASSERT_MSG(targetCurrent < 27300, "Current of %d is too high.", targetCurrent);
```
Comment (Ishadijcks) [inline march_hardware/src/IMotionCube.cpp:122]: Magic number

File: march_hardware_builder/src/robots/march4.yaml [hunk 35]
```diff
@@ -1,25 +1,25 @@
...
+#      lowVoltageNet2Current: 13
+#      highVoltageNetCurrent: 17
+#      lowVoltageState: 3
+#      highVoltageOvercurrentTrigger: 2
+#      highVoltageEnabled: 1
+#      highVoltageState: 4
+#    netDriverByteOffsets:
+#      lowVoltageNetOnOff: 4
+#      highVoltageNetOnOff: 3
+#      allHighVoltageOnOff: 2
```
Comment (Ishadijcks) [inline march_hardware_builder/src/robots/march4.yaml:22]: Why this change?

File: march_hardware_builder/src/robots/march4.yaml [hunk 36]
```diff
@@ -63,22 +66,25 @@ march4:
...
+        actuationMode: torque
         netNumber: 7
         temperatureges: {slaveIndex: 10, byteOffset: 4}
         imotioncube:
           slaveIndex: 11
           encoder: {resolution: 17, minPositionIU: 19892, maxPositionIU: 63537, zeroPositionIU: 21711, safetyMarginRad: 0.05}
 
     - right_ankle:
-        allowActuation: true
+        allowActuation: false
```
Comment (Ishadijcks) [inline march_hardware_builder/src/robots/march4.yaml:77]: Should not be changed in develop

File: march_hardware_builder/src/robots/march4.yaml [hunk 37]
```diff
@@ -63,22 +66,25 @@ march4:
...
+        actuationMode: torque
         netNumber: 4
         temperatureges: {slaveIndex: 4, byteOffset: 0}
         imotioncube:
           slaveIndex: 3
           encoder: {resolution: 12, minPositionIU: 462, maxPositionIU: 937, zeroPositionIU: 759, safetyMarginRad: 0}
 
     - left_ankle:
-        allowActuation: true
+        allowActuation: false
```
Comment (Ishadijcks) [inline march_hardware_builder/src/robots/march4.yaml:86]: Should not be changed in develop

File: march_hardware_interface/config/march4/controllers.yaml [hunk 38]
```diff
@@ -11,21 +11,26 @@ march:
...
+      allow_partial_joints_goal: true
       joints:
         - right_hip_fe
         - left_hip_fe
         - right_knee
         - left_knee
-        - right_ankle
-        - left_ankle
+#        - right_ankle
+#        - left_ankle
```
Comment (Ishadijcks) [inline march_hardware_interface/config/march4/controllers.yaml:22]: Should not be changed in develop

File: march_hardware_interface/config/march4/controllers.yaml [hunk 39]
```diff
@@ -11,21 +11,26 @@ march:
...
         type: position_controllers/JointTrajectoryController
         joints:
           - right_hip_fe
           - left_hip_fe
           - right_knee
           - left_knee
-          - right_ankle
-          - left_ankle
+#          - right_ankle
+#          - left_ankle
```
Comment (Ishadijcks) [inline march_hardware_interface/config/march4/controllers.yaml:36]: Should not be changed in develop

**Comment by Ishadijcks** [comment]
102 commits? Were they all needed?

**Comment by Ishadijcks** [comment]
Closing for now as this is not my pull request. Can be reopened once all concerns are addressed.

File: march_hardware/include/march_hardware/ActuationMode.h (hunk 5 above)
Comment (JitskedeVries) [reply march_hardware/include/march_hardware/ActuationMode.h:32]: fixed

File: march_hardware/include/march_hardware/ActuationMode.h (hunk 1 above)
Comment (JitskedeVries) [reply march_hardware/include/march_hardware/ActuationMode.h:73]: fixed

File: march_hardware/include/march_hardware/Joint.h (hunk 7 above)
Comment (JitskedeVries) [reply march_hardware/include/march_hardware/Joint.h:50]: fixed

File: march_hardware/src/EtherCAT/EthercatMaster.cpp (hunk 8 above)
Comment (JitskedeVries) [reply march_hardware/src/EtherCAT/EthercatMaster.cpp:101]: fixed

File: march_hardware/src/EtherCAT/EthercatSDO.cpp (hunk 9 above)
Comment (JitskedeVries) [reply march_hardware/src/EtherCAT/EthercatSDO.cpp:15]: fixed

File: march_hardware/src/EtherCAT/EthercatSDO.cpp (hunk 10 above)
Comment (JitskedeVries) [reply march_hardware/src/EtherCAT/EthercatSDO.cpp:21]: fixed

File: march_hardware/src/EtherCAT/EthercatSDO.cpp (hunk 11 above)
Comment (JitskedeVries) [reply march_hardware/src/EtherCAT/EthercatSDO.cpp:27]: fixed

File: march_hardware/src/IMotionCube.cpp (hunk 13 above)
Comment (JitskedeVries) [reply march_hardware/src/IMotionCube.cpp:37]: target current is removed as miso

File: march_hardware/src/IMotionCube.cpp (hunk 34 above)
Comment (JitskedeVries) [reply march_hardware/src/IMotionCube.cpp:122]: Added a comment with reasoning.

File: march_hardware/src/Joint.cpp (hunk 24 above)
Comment (JitskedeVries) [reply march_hardware/src/Joint.cpp:101]: indeed must be an int since it is in IU, changed the rest into a float

File: march_hardware/src/Joint.cpp (hunk 23 above)
Comment (JitskedeVries) [reply march_hardware/src/Joint.cpp:103]: Limits interface is working with torque mode

File: march_hardware_builder/src/robots/march4.yaml (hunk 35 above)
Comment (JitskedeVries) [reply march_hardware_builder/src/robots/march4.yaml:22]: Shouldn't be a change indeed for develop

File: march_hardware_builder/src/robots/march4.yaml (hunk 36 above)
Comment (JitskedeVries) [reply march_hardware_builder/src/robots/march4.yaml:77]: I agree

File: march_hardware_interface/config/march3/controllers.yaml (hunk 26 above)
Comment (JitskedeVries) [reply march_hardware_interface/config/march3/controllers.yaml:27]: Still working on improving the PID gains, they will be tested when we do our final pullrequest. Or wasn't that the question?

File: march_hardware_interface/launch/march4.launch [hunk 40]
```diff
@@ -9,10 +9,13 @@
           args="
                 /march/controller/joint_state
                 /march/controller/temperature_sensor
-                /march/controller/trajectory
-                /march/controller/pdb_state
+                /march/controller/trajectory_effort
             "/>
+    <!--                /march/controller/pdb_state-->
+    <!--    /march/controller/trajectory_position-->
```
Comment (JitskedeVries) [inline march_hardware_interface/launch/march4.launch:15]: Should not be changed in develop

File: march_hardware_interface/src/march_hardware_interface.cpp (hunk 29 above)
Comment (JitskedeVries) [reply march_hardware_interface/src/march_hardware_interface.cpp:163]: What do you mean?

File: march_hardware/src/IMotionCube.cpp (hunk 14 above)
Comment (JitskedeVries) [reply march_hardware/src/IMotionCube.cpp:50]: Is no comment anymore
